[v1.17] ipsec: handle tunnelled ipv6 in v1.17 leak detection

ldelossa opened the pull request:
In 38bfecaeddfd249bea0019c91047a1c5b7f02b16 code was backported to v1.17 to ensure no leaked packets would slip through when upgrading/downgrading from v1.17 to v1.18.

The leak detection mechanism failed to include checking for encapsulated IPv6 packets.

Update the leak detection bits to parse out the inner MAC header, determine the IP version, and check the inner IP protocol accordingly.

This will now ensure leaked tunnelled IPv6 traffic is encapsulated during v1.17<->v1.18 upgrade/downgrade.

```release-note
ipsec: include ipv6 in v1.18 upgrade leak detection
```

ldelossa committed:
ipsec: handle tunnelled ipv6 in v1.17 leak detection

In 38bfecaeddfd249bea0019c91047a1c5b7f02b16 code was backported to
v1.17 to ensure no leaked packets would slip through when
upgrading/downgrading from v1.17 to v1.18.

The leak detection mechanism failed to include checking for encapsulated
IPv6 packets.

Update the leak detection bits to parse out the inner MAC header,
determine the IP version, and check the inner IP protocol accordingly.

This will now ensure leaked tunnelled IPv6 traffic is encapsulated
during v1.17<->v1.18 upgrade/downgrade.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

julianwiedmann commented on bpf/lib/encrypt.h:
```diff
@@ -291,18 +294,48 @@ ipsec_maybe_redirect_to_encrypt(struct __ctx_buff *ctx, __be16 proto,
-			if (ip4_inner->protocol == IPPROTO_ESP)
+			eth_inner = (struct ethhdr *)(data + l2_inner_off);
+			switch (eth_inner->h_proto) {
+			case ETH_P_IP:
```
```suggestion
			switch (eth_inner->h_proto) {
			case bpf_htons(ETH_P_IP):
```

Should we maybe flip this PR back to draft, until you've had opportunity to add sufficient testing?

ldelossa replied:
What extent of testing are you thinking? 

I know you have the idea for an integration test here, but we didnt do this with the original backport PR.

julianwiedmann replied:
I believe a bunch of BPF tests with the packets you plan to process in this function are needed. We already have similar tests for bpf_lxc, this would simply establish equivalent coverage.

We didn't add such tests initially, but we should have. And they would have easily caught this bug, instead of relying on reviewers to spot them.

ldelossa replied:
@julianwiedmann we also have a unit test just for `ipsec_maybe_redirect_to_encrypt` which can be updated to catch this same issue. 

I'm not sure if this PR requires a full integration test of `to_netdev` or an update to the exiting unit tests that throws vxlan traffic into this function. 

Happy to do a full integration tests against to-netdev but also looking to time box this fix

ldelossa replied:
Updated with tunnel traffic tests of `ipsec_maybe_redirect_to_encrypt`.

ldelossa committed:
ipsec: add additional ipsec and vxlan helpers

Add additional helpers for vxlan encapsulated packets in the context of
IPsec.

This will be used for upgrade/downgrade leak detection between v1.17 and
v1.18.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

julianwiedmann reviewed (approved):
lgtm, thank you!
